In [3]:
import pandas as pd

In [7]:
df = pd.read_excel('metales_pesados.xlsx')

In [9]:
df.head(5)

DISTRITO ESTABLECIMIENTO               APELLIDOS_NOMBRES  EDAD  \
0  SAN PEDRO DE PILLAO          PILLAO  VARA CERVANTES MARIORET ARIANA  10.0   
1  SAN PEDRO DE PILLAO          PILLAO       VERTIZ RIVAS JHEREMY ESAU   5.0   
2  SAN PEDRO DE PILLAO          PILLAO   DIONISIO ESPINOZA MAXIMO JUAN   6.0   
3  SAN PEDRO DE PILLAO          PILLAO           RIVERA PEÑA ANDY IKER   8.0   
4  SAN PEDRO DE PILLAO          PILLAO       ROJAS FUSTER CESIA DEBORA   9.0   

  GENERO FECHA_NACIDO        SIS RESULT1  RESULT2 RESULT3 RESULT4  
0      F   2013-09-15     ACTIVO     NaN      NaN     NaN      SR  
1      M   2018-12-28     ACTIVO     NaN      NaN     NaN      SR  
2      M   2017-07-19  CANCELADO     NaN      NaN     NaN      SR  
3      M   2015-09-06     ACTIVO     NaN      NaN     NaN      SR  
4      F   2014-03-02  CANCELADO     NaN      NaN     NaN      SR

In [10]:
df_clean = df.copy()

In [11]:
df_clean.head(5)

DISTRITO ESTABLECIMIENTO               APELLIDOS_NOMBRES  EDAD  \
0  SAN PEDRO DE PILLAO          PILLAO  VARA CERVANTES MARIORET ARIANA  10.0   
1  SAN PEDRO DE PILLAO          PILLAO       VERTIZ RIVAS JHEREMY ESAU   5.0   
2  SAN PEDRO DE PILLAO          PILLAO   DIONISIO ESPINOZA MAXIMO JUAN   6.0   
3  SAN PEDRO DE PILLAO          PILLAO           RIVERA PEÑA ANDY IKER   8.0   
4  SAN PEDRO DE PILLAO          PILLAO       ROJAS FUSTER CESIA DEBORA   9.0   

  GENERO FECHA_NACIDO        SIS RESULT1  RESULT2 RESULT3 RESULT4  
0      F   2013-09-15     ACTIVO     NaN      NaN     NaN      SR  
1      M   2018-12-28     ACTIVO     NaN      NaN     NaN      SR  
2      M   2017-07-19  CANCELADO     NaN      NaN     NaN      SR  
3      M   2015-09-06     ACTIVO     NaN      NaN     NaN      SR  
4      F   2014-03-02  CANCELADO     NaN      NaN     NaN      SR

In [13]:
df_clean['EDAD'] = df_clean['EDAD'].fillna(df_clean['EDAD'].mean())

In [14]:
# Convertir las columnas a numéricas, los errores se convierten en NaN
df_clean['RESULT1'] = pd.to_numeric(df_clean['RESULT1'], errors='coerce')
df_clean['RESULT2'] = pd.to_numeric(df_clean['RESULT2'], errors='coerce')
df_clean['RESULT3'] = pd.to_numeric(df_clean['RESULT3'], errors='coerce')
df_clean['RESULT4'] = pd.to_numeric(df_clean['RESULT4'], errors='coerce')



In [15]:
# Crear la nueva columna que es la media de las otras columnas
df_clean['result_mean'] = df_clean[['RESULT1', 'RESULT2', 'RESULT3', 'RESULT4']].mean(axis=1)

In [16]:
df_clean.head(5)

DISTRITO ESTABLECIMIENTO               APELLIDOS_NOMBRES  EDAD  \
0  SAN PEDRO DE PILLAO          PILLAO  VARA CERVANTES MARIORET ARIANA  10.0   
1  SAN PEDRO DE PILLAO          PILLAO       VERTIZ RIVAS JHEREMY ESAU   5.0   
2  SAN PEDRO DE PILLAO          PILLAO   DIONISIO ESPINOZA MAXIMO JUAN   6.0   
3  SAN PEDRO DE PILLAO          PILLAO           RIVERA PEÑA ANDY IKER   8.0   
4  SAN PEDRO DE PILLAO          PILLAO       ROJAS FUSTER CESIA DEBORA   9.0   

  GENERO FECHA_NACIDO        SIS  RESULT1  RESULT2  RESULT3  RESULT4  \
0      F   2013-09-15     ACTIVO      NaN      NaN      NaN      NaN   
1      M   2018-12-28     ACTIVO      NaN      NaN      NaN      NaN   
2      M   2017-07-19  CANCELADO      NaN      NaN      NaN      NaN   
3      M   2015-09-06     ACTIVO      NaN      NaN      NaN      NaN   
4      F   2014-03-02  CANCELADO      NaN      NaN      NaN      NaN   

   result_mean  
0          NaN  
1          NaN  
2          NaN  
3          NaN  
4          NaN

In [17]:
df_clean['result_mean'] = df_clean['result_mean'].fillna(df_clean['result_mean'].mean())

In [18]:
df_clean.head(5)

DISTRITO ESTABLECIMIENTO               APELLIDOS_NOMBRES  EDAD  \
0  SAN PEDRO DE PILLAO          PILLAO  VARA CERVANTES MARIORET ARIANA  10.0   
1  SAN PEDRO DE PILLAO          PILLAO       VERTIZ RIVAS JHEREMY ESAU   5.0   
2  SAN PEDRO DE PILLAO          PILLAO   DIONISIO ESPINOZA MAXIMO JUAN   6.0   
3  SAN PEDRO DE PILLAO          PILLAO           RIVERA PEÑA ANDY IKER   8.0   
4  SAN PEDRO DE PILLAO          PILLAO       ROJAS FUSTER CESIA DEBORA   9.0   

  GENERO FECHA_NACIDO        SIS  RESULT1  RESULT2  RESULT3  RESULT4  \
0      F   2013-09-15     ACTIVO      NaN      NaN      NaN      NaN   
1      M   2018-12-28     ACTIVO      NaN      NaN      NaN      NaN   
2      M   2017-07-19  CANCELADO      NaN      NaN      NaN      NaN   
3      M   2015-09-06     ACTIVO      NaN      NaN      NaN      NaN   
4      F   2014-03-02  CANCELADO      NaN      NaN      NaN      NaN   

   result_mean  
0     5.122449  
1     5.122449  
2     5.122449  
3     5.122449  
4     5.122449

In [19]:
df_clean.columns

Index(['DISTRITO', 'ESTABLECIMIENTO', 'APELLIDOS_NOMBRES', 'EDAD', 'GENERO',
       'FECHA_NACIDO', 'SIS', 'RESULT1', 'RESULT2', 'RESULT3', 'RESULT4',
       'result_mean'],
      dtype='object')

In [20]:
df_entrenamiento = df_clean[['DISTRITO', 'ESTABLECIMIENTO', 'EDAD','GENERO','result_mean']].copy()

In [22]:
df_entrenamiento['GENERO'] = df_entrenamiento['GENERO'].replace({'F': 0, 'M': 1})

C:\Users\MABE\AppData\Local\Temp\ipykernel_13880\3359927515.py:1: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df_entrenamiento['GENERO'] = df_entrenamiento['GENERO'].replace({'F': 0, 'M': 1})


In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

# Dividir los datos en conjuntos de entrenamiento y prueba
X = df_entrenamiento[['GENERO']]
y = df_entrenamiento['result_mean'] > df_entrenamiento['result_mean'].mean()
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Entrenar el modelo
model = LogisticRegression()
model.fit(X_train, y_train)

# Evaluar el modelo
y_pred = model.predict(X_test)
print('Accuracy:', accuracy_score(y_test, y_pred))

In [25]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

# Dividir los datos en conjuntos de entrenamiento y prueba
X = df_entrenamiento[['GENERO']]  # Variable independiente (sexo)
y = df_entrenamiento['result_mean'] > df_entrenamiento['result_mean'].mean()  # Variable dependiente (mayor que la media)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Entrenar el modelo
model = LogisticRegression()
model.fit(X_train, y_train)

# Evaluar el modelo
y_pred = model.predict(X_test)
print('Accuracy:', accuracy_score(y_test, y_pred))


Accuracy: 0.7758620689655172


In [26]:
# Obtener los coeficientes del modelo
coefficients = model.coef_[0]
intercept = model.intercept_

# Imprimir los coeficientes
print("Coeficientes del modelo:")
for feature, coef in zip(X.columns, coefficients):
    print(feature, ':', coef)

# Intercepto
print("Intercepto:", intercept)


Coeficientes del modelo:
GENERO : 0.5174802717218021
Intercepto: [-1.24926571]


La fórmula para calcular la probabilidad ajustada en un modelo de regresión logística es

Donde:
P(Y=1∣X) es la probabilidad de que la variable de salida sea 1 dado el valor de la variable de entrada X.
β0 es el intercepto del modelo.
β1 es el coeficiente asociado a la variable de entrada.
X es el valor de la variable de entrada.


β0:−1.24926571 (intercepto)
β1:0.5174802717218021 (coeficiente para la variable "GENERO")


x = 1 : 0.32480
x=0 obtengo : 0.22282

Para el sexo masculino (GENERO = 1), la probabilidad ajustada de que el valor medio sea mayor que la media es aproximadamente 0.32480, lo que significa que hay alrededor del 32.48% de probabilidades de que un individuo masculino tenga un valor medio mayor que la media.
Para el sexo femenino (GENERO = 0), la probabilidad ajustada de que el valor medio sea mayor que la media es aproximadamente 0.22282, lo que significa que hay alrededor del 22.28% de probabilidades de que un individuo femenino tenga un valor medio mayor que la media.

In [35]:
# Crear variables ficticias para 'DISTRITO' y 'ESTABLECIMIENTO'
df_entrenamiento = pd.get_dummies(df_entrenamiento, columns=['DISTRITO'])

In [37]:
# Dividir los datos en conjuntos de entrenamiento y prueba
X = df_entrenamiento.drop(columns=['result_mean'])  # Variables independientes
y = df_entrenamiento['result_mean'] > df_entrenamiento['result_mean'].mean()  # Variable dependiente
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Entrenar el modelo
model = LogisticRegression()
model.fit(X_train, y_train)

# Evaluar el modelo
y_pred = model.predict(X_test)
print('Accuracy:', accuracy_score(y_test, y_pred))

Accuracy: 0.8448275862068966


In [38]:
# Obtener los coeficientes del modelo
coefficients = model.coef_[0]
intercept = model.intercept_

# Crear un DataFrame para almacenar los coeficientes y las características correspondientes
coef_df = pd.DataFrame({'Feature': X.columns, 'Coefficient': coefficients})

# Imprimir los coeficientes ordenados de mayor a menor
coef_df = coef_df.sort_values(by='Coefficient', ascending=False)



                         Feature  Coefficient
8     DISTRITO_SANTA ANA DE TUSI     0.952233
1                         GENERO     0.669078
3         ESTABLECIMIENTO_MACHIN     0.611645
2  ESTABLECIMIENTO_CHINCHE TINGO     0.552067
4       ESTABLECIMIENTO_PAMPANIA     0.340588
0                           EDAD     0.011855
5         ESTABLECIMIENTO_PILLAO    -0.352067
7   DISTRITO_SAN PEDRO DE PILLAO    -0.352067
9            DISTRITO_YANAHUANCA    -0.589337
6      ESTABLECIMIENTO_YANACOCHA    -1.141404


In [41]:
coef_df.head(20)

Feature  Coefficient
8     DISTRITO_SANTA ANA DE TUSI     0.952233
1                         GENERO     0.669078
3         ESTABLECIMIENTO_MACHIN     0.611645
2  ESTABLECIMIENTO_CHINCHE TINGO     0.552067
4       ESTABLECIMIENTO_PAMPANIA     0.340588
0                           EDAD     0.011855
5         ESTABLECIMIENTO_PILLAO    -0.352067
7   DISTRITO_SAN PEDRO DE PILLAO    -0.352067
9            DISTRITO_YANAHUANCA    -0.589337
6      ESTABLECIMIENTO_YANACOCHA    -1.141404